fastcrypto/src/vrf.rs

```diff
@@ -0,0 +1,268 @@
+    use elliptic_curve::hash2curve::{ExpandMsg, Expander};
+    use serde::{Deserialize, Serialize};
+
+    /// draft-irtf-cfrg-vrf-15 specifies suites for suite-strings 0x00-0x04 and notes that future
+    /// designs should specify a different suite_string constant, so we use 0x05 here.
+    const SUITE_STRING: [u8; 1] = [0x05]; // TODO: Update this with future releases of the draft or final version.
+
+    /// Length of challenge. Must not exceed the length of field elements which is 32 in this case and
+    /// the other suites uses half the length, so we do the same here.
+    const C_LEN: usize = 16;
+
+    /// Default hash function
+    type H = Sha512;
+
+    /// Domain separation tag used in ecvrf_encode_to_curve
+    const DST: &[u8; 43] = b"ECVRF_ristretto255_XMD:SHA-512_R255MAP_RO_5";
+
+    const COFACTOR: u64 = 8;
+
+    #[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+    pub struct ECVRFPublicKey(RistrettoPoint);
+
+    impl VRFPublicKey for ECVRFPublicKey {
+        type PrivateKey = ECVRFPrivateKey;
+    }
+
+    impl ECVRFPublicKey {
+        /// Encode the given binary string as curve point. See section 5.4.1.2 of draft-irtf-cfrg-vrf-15.
+        fn ecvrf_encode_to_curve(&self, alpha_string: &[u8]) -> RistrettoPoint {
+            // This follows section 5.4.1.2 of draft-irtf-cfrg-vrf-15 for the ristretto255 group using
+            // SHA-512. The hash-to-curve for ristretto255 follows appendix B of draft-irtf-cfrg-hash-to-curve-16.
+
+            // Compute expand_message_xmd for the given message. Note that expand_message only returns
+            // and error if the len_in_bytes and output size of the hash function is out of bounds
+            // (https://github.com/mikelodder7/hash2field/blob/cdf56a2b722aeae25b8019945afe4cccec132f25/src/expand_msg_xmd.rs#L21),
+            // so we can safely unwrap since they are constants here.
+            let mut expanded_message =
+                elliptic_curve::hash2curve::ExpandMsgXmd::<sha2::Sha512>::expand_message(
+                    &[&self.0.compress(), alpha_string],
+                    DST,
+                    H::OUTPUT_SIZE,
+                )
+                .unwrap();
+
+            let mut bytes = [0u8; H::OUTPUT_SIZE];
+            expanded_message.fill_bytes(&mut bytes);
+            RistrettoPoint::map_to_point::<H>(&bytes)
+        }
+
+        /// Implements ECVRF_validate_key which checks the validity of a public key. See section 5.4.5
+        /// of draft-irtf-cfrg-vrf-15.
+        fn valid(&self) -> bool {
+            self.0 * RistrettoScalar::from(COFACTOR) != RistrettoPoint::zero()
+        }
+    }
+
+    #[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+    pub struct ECVRFPrivateKey(RistrettoScalar);
+
+    impl VRFPrivateKey for ECVRFPrivateKey {
+        type PublicKey = ECVRFPublicKey;
+    }
+
+    impl ECVRFPrivateKey {
+        /// Generate nonce from binary string. See section 5.4.2.2. of draft-irtf-cfrg-vrf-15.
```

Comment:
why is it called a nonce and not scalar?

Reply:
To follow the terminology from the rfc draft (https://www.ietf.org/archive/id/draft-irtf-cfrg-vrf-15.html#name-ecvrf-nonce-generation), but it is a scalar so we could also call it that if you think it makes more sense.